File: backend/routes/tasks.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional
from datetime import datetime
from database import get_db
from auth import get_current_user
import models

router = APIRouter()

class TaskCreate(BaseModel):
    title: str
    description: Optional[str] = None
    project_id: int
    assignee_id: Optional[int] = None
    priority: Optional[str] = "medium"
    due_date: Optional[datetime] = None

class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    priority: Optional[str] = None
    assignee_id: Optional[int] = None
    due_date: Optional[datetime] = None

class TaskOut(BaseModel):
    id: int
    title: str
    description: Optional[str]
    status: str
    priority: str
    project_id: int
    assignee_id: Optional[int]
    created_by: int
    due_date: Optional[datetime]
    created_at: Optional[datetime]
    assignee_name: Optional[str] = None
    class Config:
        from_attributes = True

def task_to_dict(task: models.Task, db: Session):
    assignee_name = None
    if task.assignee_id:
        u = db.query(models.User).filter(models.User.id == task.assignee_id).first()
        assignee_name = u.name if u else None
    return {
        "id": task.id,
        "title": task.title,
        "description": task.description,
        "status": task.status,
        "priority": task.priority,
        "project_id": task.project_id,
        "assignee_id": task.assignee_id,
        "created_by": task.created_by,
        "due_date": task.due_date,
        "created_at": task.created_at,
        "assignee_name": assignee_name
    }

@router.post("/", status_code=201)
def create_task(
    payload: TaskCreate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):
    # Check membership
    membership = db.query(models.ProjectMember).filter(
        models.ProjectMember.project_id == payload.project_id,
        models.ProjectMember.user_id == current_user.id
    ).first()
    if not membership:
        raise HTTPException(status_code=403, detail="Not a project member")

    priority_map = {"low": models.PriorityEnum.low, "medium": models.PriorityEnum.medium, "high": models.PriorityEnum.high}
    task = models.Task(
        title=payload.title,
        description=payload.description,
        project_id=payload.project_id,
        assignee_id=payload.assignee_id,
        created_by=current_user.id,
        priority=priority_map.get(payload.priority, models.PriorityEnum.medium),
        due_date=payload.due_date
    )
    db.add(task)
    db.commit()
    db.refresh(task)
    return task_to_dict(task, db)

@router.get("/project/{project_id}")
def get_project_tasks(
    project_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):
    membership = db.query(models.ProjectMember).filter(
        models.ProjectMember.project_id == project_id,
        models.ProjectMember.user_id == current_user.id
    ).first()
    if not membership:
        raise HTTPException(status_code=403, detail="Not a project member")
    tasks = db.query(models.Task).filter(models.Task.project_id == project_id).all()
    return [task_to_dict(t, db) for t in tasks]

@router.get("/my")
def get_my_tasks(
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):
    tasks = db.query(models.Task).filter(models.Task.assignee_id == current_user.id).all()
    return [task_to_dict(t, db) for t in tasks]

@router.patch("/{task_id}")
def update_task(
    task_id: int,
    payload: TaskUpdate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):
    task = db.query(models.Task).filter(models.Task.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

    if payload.title is not None:
        task.title = payload.title
    if payload.description is not None:
        task.description = payload.description
    if payload.status is not None:
        status_map = {"todo": models.StatusEnum.todo, "in_progress": models.StatusEnum.in_progress, "done": models.StatusEnum.done}
        task.status = status_map.get(payload.status, task.status)
    if payload.priority is not None:
        priority_map = {"low": models.PriorityEnum.low, "medium": models.PriorityEnum.medium, "high": models.PriorityEnum.high}
        task.priority = priority_map.get(payload.priority, task.priority)
    if payload.assignee_id is not None:
        task.assignee_id = payload.assignee_id
    if payload.due_date is not None:
        task.due_date = payload.due_date

    db.commit()
    db.refresh(task)
    return task_to_dict(task, db)

@router.delete("/{task_id}", status_code=204)
def delete_task(
    task_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):
    task = db.query(models.Task).filter(models.Task.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    membership = db.query(models.ProjectMember).filter(
        models.ProjectMember.project_id == task.project_id,
        models.ProjectMember.user_id == current_user.id
    ).first()
    if not membership:
        raise HTTPException(status_code=403, detail="Not authorized")
    db.delete(task)
    db.commit()

File: backend/src/routes/tasks.js
const express = require('express');
const Task = require('../models/Task');
const Project = require('../models/Project');
const { auth } = require('../middleware/auth');
const router = express.Router();

// Helper to check project access
const checkProjectAccess = async (projectId, userId) => {
  const project = await Project.findById(projectId);
  if (!project) return null;
  const hasAccess = project.owner.equals(userId) || project.members.some(m => m.equals(userId));
  return hasAccess ? project : null;
};

// GET tasks for a project
router.get('/project/:projectId', auth, async (req, res) => {
  try {
    const project = await checkProjectAccess(req.params.projectId, req.user._id);
    if (!project) return res.status(403).json({ message: 'Access denied or project not found' });

    const tasks = await Task.find({ project: req.params.projectId })
      .populate('assignedTo', 'name email')
      .populate('createdBy', 'name email')
      .sort({ createdAt: -1 });
    res.json(tasks);
  } catch (err) {
    res.status(500).json({ message: 'Server error', error: err.message });
  }
});

// GET dashboard stats
router.get('/dashboard', auth, async (req, res) => {
  try {
    const projects = await Project.find({
      $or: [{ owner: req.user._id }, { members: req.user._id }]
    });
    const projectIds = projects.map(p => p._id);

    const now = new Date();
    const [total, todo, inProgress, done, overdue] = await Promise.all([
      Task.countDocuments({ project: { $in: projectIds } }),
      Task.countDocuments({ project: { $in: projectIds }, status: 'todo' }),
      Task.countDocuments({ project: { $in: projectIds }, status: 'in-progress' }),
      Task.countDocuments({ project: { $in: projectIds }, status: 'done' }),
      Task.countDocuments({ project: { $in: projectIds }, status: { $ne: 'done' }, dueDate: { $lt: now } }),
    ]);

    const recentTasks = await Task.find({ project: { $in: projectIds } })
      .populate('project', 'name')
      .populate('assignedTo', 'name')
      .sort({ createdAt: -1 })
      .limit(5);

    res.json({ total, todo, inProgress, done, overdue, recentTasks, projectCount: projects.length });
  } catch (err) {
    res.status(500).json({ message: 'Server error', error: err.message });
  }
});

// POST create task
router.post('/', auth, async (req, res) => {
  try {
    const { title, description, projectId, assignedTo, priority, dueDate } = req.body;
    if (!title || !projectId) return res.status(400).json({ message: 'Title and project required' });

    const project = await checkProjectAccess(projectId, req.user._id);
    if (!project) return res.status(403).json({ message: 'Access denied or project not found' });

    const task = new Task({
      title, description, project: projectId,
      assignedTo: assignedTo || null,
      createdBy: req.user._id,
      priority: priority || 'medium',
      dueDate: dueDate || null,
    });
    await task.save();
    await task.populate('assignedTo', 'name email');
    await task.populate('createdBy', 'name email');
    res.status(201).json(task);
  } catch (err) {
    res.status(500).json({ message: 'Server error', error: err.message });
  }
});

// PUT update task
router.put('/:id', auth, async (req, res) => {
  try {
    const task = await Task.findById(req.params.id);
    if (!task) return res.status(404).json({ message: 'Task not found' });

    const project = await checkProjectAccess(task.project, req.user._id);
    if (!project) return res.status(403).json({ message: 'Access denied' });

    const { title, description, status, assignedTo, priority, dueDate } = req.body;
    if (title) task.title = title;
    if (description !== undefined) task.description = description;
    if (status) task.status = status;
    if (assignedTo !== undefined) task.assignedTo = assignedTo || null;
    if (priority) task.priority = priority;
    if (dueDate !== undefined) task.dueDate = dueDate || null;

    await task.save();
    await task.populate('assignedTo', 'name email');
    await task.populate('createdBy', 'name email');
    res.json(task);
  } catch (err) {
    res.status(500).json({ message: 'Server error', error: err.message });
  }
});

// DELETE task
router.delete('/:id', auth, async (req, res) => {
  try {
    const task = await Task.findById(req.params.id);
    if (!task) return res.status(404).json({ message: 'Task not found' });

    const project = await checkProjectAccess(task.project, req.user._id);
    if (!project) return res.status(403).json({ message: 'Access denied' });

    if (req.user.role !== 'admin' && !task.createdBy.equals(req.user._id))
      return res.status(403).json({ message: 'Only task creator or admin can delete' });

    await task.deleteOne();
    res.json({ message: 'Task deleted' });
  } catch (err) {
    res.status(500).json({ message: 'Server error', error: err.message });
  }
});

module.exports = router;
